server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileTransactionLog.java

```diff
@@ -0,0 +1,385 @@
+      List<DatafileTransaction> transactions;
+      StoredTabletFile[] finalFiles;
+      boolean consistent;
+      do {
+        long updateCount = this.updateCount;
+        initialTs = this.initialTs;
+        initialFiles = this.initialFiles;
+        finalFiles = this.finalFiles;
+        transactions = this.tabletLog.toList();
+        consistent =
+            (updateCount == this.updateCount && updateCount == this.tabletLog.getUpdateCount());
+      } while (!consistent);
+
+      // now clear out the log if requested
+      if (clear) {
+        clear();
+      }
+
+      // now we can build our dump string
+      StringBuilder builder = new StringBuilder();
+      SimpleDateFormat format = new SimpleDateFormat(DATE_FORMAT);
+      String initialDate = format.format(Date.from(Instant.ofEpochMilli(initialTs)));
+
+      builder.append(String.format("%s: Initial files for %s : %s", initialDate, extent,
+          new TreeSet<>(Arrays.asList(initialFiles))));
+      transactions.stream().forEach(t -> builder.append('\n').append(t.toString(format)));
+      if (!transactions.isEmpty()) {
+        builder.append("\nFinal files: ").append(new TreeSet<>(Arrays.asList(finalFiles)));
+      }
+
+      return builder.toString();
+    }
+  }
+
+  /**
+   * The max size of the log as derived from the configuration
+   */
+  private static class MaxLogSize {
+    private final int maxSize;
+
+    public MaxLogSize(AccumuloConfiguration config) {
+      maxSize = config.getCount(Property.TABLE_OPERATION_LOG_MAX_SIZE);
+    }
+
+    public int getMaxSize() {
+      return maxSize;
+    }
+  }
+
+  /**
+   * A simple implementation of a ring buffer
+   */
+  public static class Ring<T> {
+    private Object[] ring;
+    private volatile int first;
+    private volatile int last;
+
+    public Ring(int size) {
+      ring = new Object[size];
+      first = 0;
+      last = -1;
+    }
+
+    @SuppressWarnings("unchecked")
+    public T add(T object) {
+      Object removed = null;
+      if (ring.length > 0) {
+        if (size() == ring.length) {
+          removed = ring[first++ % ring.length];
+        }
+        int newEnd = last + 1;
+        ring[newEnd % ring.length] = object;
+        last = newEnd;
+      } else {
+        removed = object;
+      }
+      return (T) removed;
+    }
+
+    public int size() {
+      return last - first + 1;
+    }
+
+    public int capacity() {
+      return ring.length;
+    }
+
+    public boolean isEmpty() {
+      return last < first;
+    }
+
+    public void clear() {
+      last = -1;
+      first = 0;
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<T> toList() {
+      if (isEmpty()) {
+        return Collections.emptyList();
+      }
+
+      Object[] data = null;
+      boolean consistent = false;
+      do {
+        long updateCount = getUpdateCount();
+        int lastPos = last;
+        int firstPos = first;
+        data = new Object[lastPos - firstPos + 1];
```

Comment:
It seems like this method is trying to be tolerant of multiple threads with the consistent loop.  However there are still race conditions.   It possible that between reading `last` and reading `first` that `first` is incremented by other threads causing `lastPos - firstPos + 1` to evaluate to a negative number which would cause an exception in the creation of the array.

I think this class could be like hashset or arraylist, assume the caller is a single thread or the caller takes care of sync

Reply:
Why are we creating our own ring buffer instead of using something that's mature and well tested (https://lmax-exchange.github.io/disruptor/user-guide/index.html)?

Reply:
Those other ring implementations are way too much for what I am trying to do here and does not supply some of the functionality I was looking for such as returning the removed object when adding a new object to the ring.  Trying to KISS.
In terms of thread safety, I was not going for that.  All of the modification calls are being done withing a Tablet lock, so we are good from that standpoint.  Then only thing I needed to ensure was that the toList method on the Ring returned a consistent view at the time it is called.  I think I achieved that.